storage2/src/state.rs

```diff
@@ -0,0 +1,47 @@
+use std::collections::HashMap;
+
+use jmt::{
+    storage::{TreeReader, TreeWriter},
+    JellyfishMerkleTree,
+};
+
+mod transaction;
+use transaction::Transaction as StateTransaction;
+
+/// State is a lightweight copy-on-write fork of the chain state,
+/// implemented as a RYW cache over a pinned JMT version.
+pub(crate) struct State<'a, R: TreeReader + TreeWriter> {
```

Comment:
This struct shouldn't be `pub(crate)`, since it's the public API we want people to use. We also want it to have a `'static` lifetime (i.e., no lifetime bounds), because it should be possible for it to live as long as the user wants (and maintain a consistent view of the pinned snapshot in that time), and we don't want it to have any trait bounds required (we're not generic over the storage, we are the storage).